### PR TITLE
Extension of Outlier Detection Comparison in 3-D dataset

### PR DESCRIPTION
**Pull request by MaiRajborirug**
[Plot_anomaly_comparison_3d.zip](https://github.com/NeuroDataDesign/scikit-learn/files/3882080/Plot_anomaly_comparison_3d.zip)

According to  [issue#5](https://github.com/NeuroDataDesign/scikit-learn/issues/5)  I want to PR the `Plot_anomaly_comparison_3D_SKlearn` for SKlearn and `Plot_anomaly_comparison_3D_USPORF` for USPORF. The USPORF file will have two additional algorithms, which are [`Extended Isolation Forest`](hhttps://github.com/sahandha/eif/blob/master/eif_old.py) and [`USPORF`](https://github.com/neurodata/SPORF/blob/staging/Python/rerf/urerf.py).

### CODE REVIEW

File: plot_anomaly_comparison-3D_SKlearn.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    X = np.append(X, X_noise, axis=1)
+    AUC_row=np.array([])
+    for name, algorithm in anomaly_algorithms: 
+        t0 = time.time()
+        algorithm.fit(X)
+        t1 = time.time()
+
+        # fit the data and tag outliers
+        if name == "Local Outlier Factor":
+            y_pred = algorithm.fit_predict(X)
```
Comment (bdpedigo) [inline plot_anomaly_comparison-3D_SKlearn.py]: I'm confused do the others not implement `fit_predict`, or would `fit().predict()` not work here?

**Comment by bdpedigo** [comment]
@MaiRajborirug nice progress on the notebook. A few comments after a quick look through it: 
 - There is a lot going on in this notebook; we need to condense it into a clear notebook and right now the code is a little all over the place. A few things to clean up the code: 
     - Use a code formatter like `black`, right now it's hard to read simply because of whitespace and line lengths being wrong
     - I wonder if it would be more clear to have each simulation setting be a function, thoughts? 
 - I don't think you are using AUC properly. I would closely read [this](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html) noting the order of the parameters, and the fact that you aren't supposed to input binary classifications but rather, posterior estimates. If you can't get posteriors out of a single model for some of these classifiers you may have to sweep over the values of some parameter manually. 
 - I could be wrong but I believe jovo was asking to see the actual ROC curves (in addition to your simulation plot). Personally I think it makes sense to keep accuracy or whatever you were plotting before on those plots of the simulations. The main point is that you are choosing some parameters arbitrarily so an ROC curve will allow you to compare more fairly over a range of parameters instead. 
- Nice work on the visualization, it's really cool. I wonder whether it would be worth writing functions for some of your simulations and storing them somewhere other people on your team could use? Like maybe in our fork of sklearn we could write some additional simulations like the ones that they have. Just a thought.

File: plot_anomaly_comparison-3D_SKlearn.py (hunk 1 above)
Comment (MaiRajborirug) [reply plot_anomaly_comparison-3D_SKlearn.py]: From what I know, `Local Outlier Factor` (LOF) doesn't have score_sample and decision_function method. Thus, this might be the reason they use a different function, `fit_predict`, to label the inliers/outliers while other algorithm use `fit().predict()`. Here is an [example](https://scikit-learn.org/stable/auto_examples/neighbors/plot_lof_outlier_detection.html).

**Comment by MaiRajborirug** [comment]
@bdpedigo 

Here are my updates on my PR [`commit 81ba3f4`](https://github.com/NeuroDataDesign/scikit-learn/commit/81ba3f4408fc3f025ecedf7aad8aa4e28458317d)
- Implement AUC-ROC curve and accuracy to measure the algorithms performance
- Clean up the tutorial in both .py and .ipynb
- Describe how to generate data and discuss result on the note
- Use `black` format

File: examples/plot_anomaly_comparison-3D_SKlearn-Copy2.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
+"""
+============================================================================
+Comparing anomaly detection algorithms for outlier detection on 3D toy datasets
+============================================================================
+
+
+Brief overview
+---------------------------------
+This example is an extension of anomaly detection comparison from ``sckit-learn 2.7.1``. These experiment applied six algorithms on different 3D toy dataset. Four algorithms are in sklearn, which are ``sklearn.covariance.EllipticEnvelope``, ``sklearn.svm.OneClassSVM``, ``sklearn.ensemble.IsolationForest``, and ``sklearn.neighbors.LocalOutlierFactor``. The other two algorithms are not from sklearn and are needed to install separately. They are ``Extended Isolation Forest`` and ``USPORF``. The outlier detecting performance is measure using. Accuracy score and ROC graph from ``sklearn.metrics.roc_auc_score``.
```
Comment (bdpedigo) [inline examples/plot_anomaly_comparison-3D_SKlearn-Copy2.py]: this line is way too long

File: examples/plot_anomaly_comparison-3D_SKlearn-Copy2.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+---------------------------------
+Each algorithm will predict the outliers and inliers. There are two performance measurement in this example.
+    
+  * ``sklearn.metrics.accuracy_score``
+  * ``sklearn.metrics.roc_auc_score``
+      * ``sklearn.metrics.roc_curve``
+
+Result and disscussion 
+---------------------------------
+``sklearn.covariance.EllipticEnvelope`` shows best result in high ``d_noise = 10``. However, since robust covariance create a ellptical envelope for inliers, we need more test on a misaligned data
```
Comment (bdpedigo) [inline examples/plot_anomaly_comparison-3D_SKlearn-Copy2.py]: all lines should be <88 chars

Comment (MaiRajborirug) [reply examples/plot_anomaly_comparison-3D_SKlearn-Copy2.py]: I will make a change accordingly in the most [file changes](https://github.com/NeuroDataDesign/scikit-learn/pull/9/files)

**Comment by MaiRajborirug** [comment]
Hi @bdpedigo  I already adjust the comment line to have < 88 characters. Could you take a look at them again? The most recent files is in [2 file changes](https://github.com/NeuroDataDesign/scikit-learn/pull/9/files)

**Comment by bdpedigo** [comment]
- I don't know what you mean by scikit-learn 2.7.1, the most recent version is 0.22
- don't need the text below "Import Packages"
- I would explain `d_noise` more, many people will not be familiar with it
- "ratio of outliers to all sample (uniform distribution)" I don't know why you say uniform distribution here
- "The data has three information dimension" -> Each dataset has three signal dimensions
- "Linear line" -> Linear
- I think you mix up notation in gaussian mixture explanation, what is "w"
- put space between (0.2, 0.2, 0.2) and (mu, mu, mu)
- how can the pi's not sum to 1?
- your noise is not uniform, it is gaussian, correct?
- seems weird that you made a function for one simulation and none of the others, maybe be consistent for all of them, might as well make them functions I think
- have you played with different values of gamma for SVM, or at least, letting gamma be 'scale' or 'auto'?
- capitalize auc in figure
- I'd like to see you explain what you are varying for the AUC curves. Don't really get it from the code 
- Make the text bold and bigger at the top, and not red
- Main comment is I still think these curves and results could look fairly different depending on the hyperparameters of the algorithms used. Maybe don't have to look into that for now but something to keep in mind
Despite all of the minor comments above, notebook looks good and your plots have come a long way. Nice job!

**Comment by MaiRajborirug** [comment]
> * I don't know what you mean by scikit-learn 2.7.1, the most recent version is 0.22
> * don't need the text below "Import Packages"
> * I would explain `d_noise` more, many people will not be familiar with it
> * "ratio of outliers to all sample (uniform distribution)" I don't know why you say uniform distribution here
> * "The data has three information dimension" -> Each dataset has three signal dimensions
> * "Linear line" -> Linear
> * I think you mix up notation in gaussian mixture explanation, what is "w"
> * put space between (0.2, 0.2, 0.2) and (mu, mu, mu)
> * how can the pi's not sum to 1?
> * your noise is not uniform, it is gaussian, correct?
> * seems weird that you made a function for one simulation and none of the others, maybe be consistent for all of them, might as well make them functions I think
> * have you played with different values of gamma for SVM, or at least, letting gamma be 'scale' or 'auto'?
> * capitalize auc in figure
> * I'd like to see you explain what you are varying for the AUC curves. Don't really get it from the code
> * Make the text bold and bigger at the top, and not red
> * Main comment is I still think these curves and results could look fairly different depending on the hyperparameters of the algorithms used. Maybe don't have to look into that for now but something to keep in mind
>   Despite all of the minor comments above, notebook looks good and your plots have come a long way. Nice job!

**Comment by MaiRajborirug** [comment]
Hi @bdpedigo, I re-uploaded the files according the your suggestion. The bullet points that I crossed means I followed them straight forward. Some bullet points I added some comments to answer your questions.

Please let me know whether or not these tutorials are ready to merge.

- ~~I don't know what you mean by scikit-learn 2.7.1, the most recent version is 0.22~~
  - Updated in the __Brief overview__ section
- ~~don't need the text below "Import Packages"~~
- ~~I would explain d_noise more, many people will not be familiar with it~~
  - Updated in __Data simulation__ section
- "ratio of outliers to all sample (uniform distribution)" I don't know why you say uniform distribution here
  - Update in __Data simulation__ section. I meant to say that I generate outliers using uniform distribution because I saw many outlier detection example also use uniform distribution. My guess is to make each outlier has small to no correlation with other data point.
- ~~"The data has three information dimension" -> Each dataset has three signal dimensions~~
- ~~"Linear line" -> Linear~~
- ~~I think you mix up notation in gaussian mixture explanation, what is "w"~~
  - Updated in __Data simulation__ section. I removed $w$ and make the note more clear about how to generate the three Gaussain blobs.
- ~~put space between (0.2, 0.2, 0.2) and (mu, mu, mu)~~
- ~~how can the pi's not sum to 1?~~
- ~~your noise is not uniform, it is gaussian, correct?~~
- ~~seems weird that you made a function for one simulation and none of the others, maybe be consistent for all of them, might as well make them functions I think~~
- ~~have you played with different values of gamma for SVM, or at least, letting gamma be 'scale' or 'auto'?~~
  - I tried both and decide to use 'scale' because it have a better outlier prediction
- ~~capitalize auc in figure~~
- I'd like to see you explain what you are varying for the AUC curves. Don't really get it from the code
- ~~Make the text bold and bigger at the top, and not red~~
- Main comment is I still think these curves and results could look fairly different depending on the hyperparameters of the algorithms used. Maybe don't have to look into that for now but something to keep in mind
  - I certainly agreed! This will be a good project if I can work with Jenny on hyper-parameter tuning

**Comment by bdpedigo** [comment]
> - I'd like to see you explain what you are varying for the AUC curves. Don't really get it from the code
> - Main comment is I still think these curves and results could look fairly different depending on the hyperparameters of the algorithms used. Maybe don't have to look into that for now but something to keep in mind
    - I certainly agreed! This will be a good project if I can work with Jenny on hyper-parameter tuning

I think these remaining comments are important and we should keep them in mind next semester. 

However, very well done @MaiRajborirug! You are done!
